_posts/2017-10-04-scalamacros.md

```diff
@@ -0,0 +1,299 @@
+are used.
+Nevertheless, we believe that v3 represents a significant enough improvement to
+forgive this restriction, which is admittedly an inconvenience for macro authors
+but arguably not a blocker for adoption.
+
+## Next steps
+
+As the history above shows, establishing a stable macro system for Scala is a
+large undertaking.
+It has taken many years to reach where we are today, involving a collaboration
+between many different parties.
+We still have a long way to go to reach the level of expressiveness, robustness
+and simplicity that we seek in a stable macro system.
+
+Here below is a rough estimated roadmap for macros v3
+
+- in Scala 2.12, we experiment with compiler plugins
+- in Dotty, Liu Fengyun at EPFL will work on integrations
+- in IntelliJ, Mikhail Mutcianko from the Scala Plugin team at Jetbrains will
+  work on integrations
+- in Scala 2.13, we continue to experiment via compiler plugins and compiler
+  feature flags in later minor releases
+- in Scala 2.14 macros v3 become no longer “experimental” and scala.reflect is
+  deprecated
+
+Following the recommendation of the Scala Center Advisory Board, the work on
+macros v3 will be an iterative processes between
+
+-  implementing macro features that have been approved for inclusion into
+   macros v3
+-  gathering feedback from the community on what macro features merit inclusion
+   in macros v3
+
+As for the first part, we immediately begin development to support a limited
+subset of blackbox def macros.
+
+### Blackbox def macros
+
+Blackbox def macros have non-controversially proven themselves to be invaluable
+for the Scala community.
+Examples of blackbox def macros include: `CaseClass.fieldNames` above,
+ScalaTest `assert` and Play/Circe JSON automatic readers/writers.
+Blackbox def macros share the following attributes:
+
+- they're invoked at compile-time instead of runtime
+- they can query the compiler for semantic information such as types and symbols.
+- they faithfully respect their declared type signatures, making their
+  implementation irrelevant to understand their behavior. From the end user's
+  perspective, they look and behave much like regular Scala methods excluding
+  the ability to emit compiler errors/warnings
+
+These attributes of blackbox def macros enable them to mix naturally into
+Scala codebases and make them prime for inclusion in the Scala Language
+Specification.
+
+### SIP proposal
+
+Alongside prototyping preliminary support for a limited set of blackbox def
+macros, we will immediately begin preparing a SIP proposal to include macros
+v3 into the Scala Language Specification.
+We plan to address the valuable reviews made to [SIP-29] on inline/meta in a
+new proposal, so that SIP-29 can be rejected.
+In addition, we will document how we we aim to solve hygiene using an
+innovation discovered by the collaboration of Liu Fengyun and Eugene Burmako.
+Hygiene plagues most macro tutorials and was previously considered to be out of
```

Comment:
Why include this point here? It seems floating/irrelevant.

Reply:
Removed.